rust/src/accountant.rs

```diff
@@ -0,0 +1,172 @@
+use serde_json::json;
+
+static DEFAULT_TOPIC_NAME: &str = "shared-resources-usage";
+/// This is the entry point for the library. It is in most cases
+/// everything you need to instrument your application.
+///
+/// The UsageAccountant needs a `Producer` and a `UsageAccumulator`.
+/// Data is stored in the accumulator and, periodically, it is
+/// flushed into the Kafka topic via a consumer.
+///
+/// Accumulating data locally is critical to reduce the performance impact
+/// of this library to a minimum and reduce the amount of Kafka messages.
+/// This means that this structure should be instantiated rarely.
+/// Possibly only once per application (or per thread).
+///
+/// Avoid creating a UsageAccountant every time some data needs to
+/// be recorded.
+
+pub struct UsageAccountant<'a> {
+    accumulator: UsageAccumulator,
+    producer: Box<dyn Producer + 'a>,
+    topic: String,
+}
+
+impl<'a> UsageAccountant<'a> {
+    /// Instantiates a UsageAccountant from a Kafka config object.
+    /// This initialization method lets the `UsageAccountant` create
+    /// the producer and own it.
+    ///
+    /// You should very rarely change topic name or granularity.
+    pub fn new(
+        producer_config: KafkaConfig,
+        topic_name: Option<&str>,
+        granularity: Option<Duration>,
+    ) -> UsageAccountant<'a> {
+        UsageAccountant::new_with_producer(
+            Box::new(KafkaProducer::new(producer_config)),
+            topic_name,
+            granularity,
+        )
+    }
+
+    /// Leaves the responsibility to provide a producer to the
+    /// client. Most of the times you should not need to use this.
+    pub fn new_with_producer(
+        producer: Box<dyn Producer + 'a>,
+        topic_name: Option<&str>,
+        granularity: Option<Duration>,
+    ) -> UsageAccountant<'a> {
+        let topic = topic_name.unwrap_or(DEFAULT_TOPIC_NAME).to_string();
+
+        UsageAccountant {
+            accumulator: UsageAccumulator::new(granularity),
+            producer,
+            topic,
+        }
+    }
+
+    /// Records an mount of usage for a resource, and app_feature.
+    ///
+    /// It flushes the batch if that is ready to be flushed.
+    /// The timestamp used is the system timestamp.
+    pub fn record(
+        &mut self,
+        resource_id: &str,
+        app_feature: &str,
+        amount: u64,
+        unit: UsageUnit,
+    ) -> Result<(), ClientError> {
+        let current_time = Local::now();
+        self.accumulator
+            .record(current_time, resource_id, app_feature, amount, unit);
+        if self.accumulator.should_flush(current_time) {
+            self.flush()?;
+        }
+        Ok(())
+    }
+
+    /// Forces a flush of the existing batch.
+    ///
+    /// This method should be called manually only when the application
+    /// is about to shut down or the `UsageAccountant` is about to be
+    /// destroyed.
+    pub fn flush(&mut self) -> Result<(), ClientError> {
```

Comment:
Can it be done in Drop so the user doesn't have to remember to call it manually?

Reply:
done